app/src/wpm.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#include <zmk/event_manager.h>
+#include <zmk/events/wpm_state_changed.h>
+#include <zmk/events/keycode_state_changed.h>
+
+#include <zmk/wpm.h>
+
+#define WPM_UPDATE_INTERVAL_SECONDS 1
+#define WPM_RESET_INTERVAL_SECONDS 5
+
+// See https://en.wikipedia.org/wiki/Words_per_minute
+// "Since the length or duration of words is clearly variable, for the purpose of measurement of
+// text entry, the definition of each "word" is often standardized to be five characters or
+// keystrokes long in English"
+#define CHARS_PER_WORD 5.0
+
+static uint8_t wpm_state;
+static uint8_t wpm_update_counter;
+static uint32_t key_pressed_count;
+
+int zmk_wpm_get_state() { return wpm_state; }
+
+int wpm_event_listener(const struct zmk_event_header *eh) {
+
+    LOG_DBG("Got an event");
+    if (is_keycode_state_changed(eh)) {
+        struct keycode_state_changed *ev = cast_keycode_state_changed(eh);
+        LOG_DBG("Its a keycode: %d %d", ev->keycode, ev->state);
+        // count only key up events
+        if (!ev->state) {
+            key_pressed_count++;
+            LOG_DBG("It's a key up - new count %d", key_pressed_count);
+        }
+    }
+    return 0;
+}
+
+void wpm_work_handler(struct k_work *work) {
+    wpm_update_counter++;
+    wpm_state = (key_pressed_count / CHARS_PER_WORD) /
+                (wpm_update_counter * WPM_UPDATE_INTERVAL_SECONDS / 60.0);
+
+    if (wpm_update_counter == WPM_RESET_INTERVAL_SECONDS) {
```

Comment:
Maybe change this into >= to make sure there's no chance of missing the interval

Reply:
Will do